File: scripts/skills-sync.ts
// SPDX-License-Identifier: Apache-2.0
// #!/usr/bin/env bun
/**
 * skills-sync.ts — pull a remote SKILL.md catalog into ./.claude/skills/.
 *
 * Usage:
 *     bun run scripts/skills-sync.ts <github-owner>/<repo>[#<ref>]
 *
 * Examples:
 *     bun run scripts/skills-sync.ts vercel-labs/agent-skills
 *     bun run scripts/skills-sync.ts anthropics/skills#main
 *
 * Strategy:
 *   1. Shallow-clone the repo into a temp directory (bun shell + git).
 *   2. Discover every `SKILL.md` (recursive) and copy its containing folder
 *      into `.claude/skills/<skill-name>/` — `<skill-name>` is taken from
 *      the SKILL.md YAML frontmatter `name:` field, falling back to the
 *      directory basename.
 *   3. Skip skills that already exist locally (no overwrite without --force).
 *   4. Print a summary table.
 *
 * No Node dependency — uses Bun stdlib (Bun.file, Bun.$, fs/promises).
 */

import { $ } from "bun";
import { mkdir, readdir, rm, stat, copyFile, readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join, basename, dirname, relative } from "node:path";
import { tmpdir } from "node:os";

const REPO_ROOT = join(import.meta.dir, "..");
const TARGET_ROOT = join(REPO_ROOT, ".claude", "skills");

const FORCE = process.argv.includes("--force");
const args = process.argv.slice(2).filter((a) => !a.startsWith("--"));
const source = args[0];

if (!source) {
    console.error("usage: bun run scripts/skills-sync.ts <owner>/<repo>[#<ref>] [--force]");
    process.exit(64);
}

const [repoSpec, ref = "HEAD"] = source.split("#");
if (!repoSpec.includes("/")) {
    console.error(`Invalid source "${source}" — expected <owner>/<repo>.`);
    process.exit(64);
}

const url = `https://github.com/${repoSpec}.git`;
const workDir = await Bun.file(join(tmpdir(), `skills-sync-${Date.now()}`)).name;
await mkdir(workDir, { recursive: true });

console.log(`▸ Cloning ${url}#${ref} (shallow) ...`);
const cloneResult = await $`git clone --depth 1 --branch ${ref === "HEAD" ? "main" : ref} ${url} ${workDir}`
    .nothrow();
if (cloneResult.exitCode !== 0) {
    // fallback: clone default branch
    await rm(workDir, { recursive: true, force: true });
    await mkdir(workDir, { recursive: true });
    await $`git clone --depth 1 ${url} ${workDir}`;
}

console.log("▸ Discovering SKILL.md files ...");
const candidates: string[] = [];
async function walk(dir: string) {
    for (const entry of await readdir(dir, { withFileTypes: true })) {
        if (entry.name === ".git" || entry.name === "node_modules") continue;
        const full = join(dir, entry.name);
        if (entry.isDirectory()) await walk(full);
        else if (entry.isFile() && entry.name === "SKILL.md") candidates.push(full);
    }
}
await walk(workDir);
console.log(`  found ${candidates.length} skill(s).`);

await mkdir(TARGET_ROOT, { recursive: true });
const report: { name: string; status: "installed" | "skipped" | "overwritten"; from: string }[] = [];

for (const skillMd of candidates) {
    const text = await readFile(skillMd, "utf8");
    const frontmatter = text.match(/^---\s*\n([\s\S]*?)\n---/);
    let name = basename(dirname(skillMd));
    if (frontmatter) {
        const m = frontmatter[1].match(/^name:\s*(.+?)\s*$/m);
        if (m) name = m[1].trim();
    }
    const dest = join(TARGET_ROOT, name);
    const srcDir = dirname(skillMd);
    if (existsSync(dest) && !FORCE) {
        report.push({ name, status: "skipped", from: relative(workDir, srcDir) });
        continue;
    }
    if (existsSync(dest)) await rm(dest, { recursive: true, force: true });
    await copyDir(srcDir, dest);
    report.push({
        name,
        status: existsSync(dest) && FORCE ? "overwritten" : "installed",
        from: relative(workDir, srcDir),
    });
}

await rm(workDir, { recursive: true, force: true });

console.log("");
console.log("Summary");
console.log("-------");
for (const r of report) {
    const tag = r.status === "skipped" ? "SKIP" : r.status === "overwritten" ? "REPL" : "NEW ";
    console.log(`  [${tag}] ${r.name.padEnd(40)}  ← ${r.from}`);
}
console.log("");
console.log(`Done. ${report.filter((r) => r.status !== "skipped").length} skill(s) written to ${TARGET_ROOT}.`);

async function copyDir(src: string, dest: string): Promise<void> {
    await mkdir(dest, { recursive: true });
    for (const entry of await readdir(src, { withFileTypes: true })) {
        const s = join(src, entry.name);
        const d = join(dest, entry.name);
        if (entry.isDirectory()) await copyDir(s, d);
        else if (entry.isFile()) await copyFile(s, d);
    }
}

File: crates/aphrody-xtask/src/skills_sync.rs
// skills_sync.rs — pull a remote SKILL.md catalog into ./.claude/skills/.
//
// Rust port of scripts/skills-sync.ts (Bun). Same CLI surface:
//
//     cargo xtask skills-sync <owner>/<repo>[#<ref>] [--force]
//
// Strategy:
//   1. Shallow-clone the repo into a temp directory (system `git`).
//   2. Walk the clone, find every `SKILL.md`.
//   3. Parse YAML frontmatter for `name:` (falls back to dir basename).
//   4. Copy the containing folder into `.claude/skills/<name>/`
//      unless it already exists (use `--force` to overwrite).
//   5. Print a summary table to stdout.

use std::{
    env, fs,
    path::{Path, PathBuf},
    process::Command,
};

use anyhow::{anyhow, bail, Context, Result};
use clap::Args as ClapArgs;
use serde::Deserialize;
use walkdir::WalkDir;

#[derive(Debug, ClapArgs)]
pub(crate) struct Args {
    /// Remote source as `<owner>/<repo>[#<ref>]`.
    /// Examples: `vercel-labs/agent-skills`, `anthropics/skills#main`.
    source: String,

    /// Overwrite existing skills locally.
    #[arg(long)]
    force: bool,
}

#[derive(Debug, Deserialize)]
struct Frontmatter {
    name: Option<String>,
}

enum Status {
    Installed,
    Skipped,
    Overwritten,
}

impl Status {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Installed => "installed",
            Self::Skipped => "skipped",
            Self::Overwritten => "overwritten",
        }
    }
}

pub(crate) fn run(args: Args) -> Result<()> {
    let (repo_spec, refname) = split_source(&args.source)?;
    let url = format!("https://github.com/{repo_spec}.git");

    let workdir = tempfile::tempdir().context("create temp dir for clone")?;
    let workdir_path = workdir.path();

    println!("\u{25b8} Cloning {url}#{refname} (shallow) ...");
    let branch_for_initial = if refname == "HEAD" { "main" } else { &refname };
    let initial = Command::new("git")
        .args(["clone", "--depth", "1", "--branch", branch_for_initial])
        .arg(&url)
        .arg(workdir_path)
        .status()
        .context("spawn git for initial clone")?;
    if !initial.success() {
        // Fallback: default branch.
        for entry in fs::read_dir(workdir_path).context("read tmp dir for cleanup")? {
            let entry = entry?;
            let p = entry.path();
            if p.is_dir() {
                fs::remove_dir_all(&p).ok();
            } else {
                fs::remove_file(&p).ok();
            }
        }
        let fallback = Command::new("git")
            .args(["clone", "--depth", "1"])
            .arg(&url)
            .arg(workdir_path)
            .status()
            .context("spawn git for fallback clone")?;
        if !fallback.success() {
            bail!("git clone failed for {url}");
        }
    }

    println!("\u{25b8} Discovering SKILL.md files ...");
    let mut candidates = Vec::new();
    for entry in WalkDir::new(workdir_path)
        .into_iter()
        .filter_entry(|e| {
            let name = e.file_name().to_string_lossy();
            name != ".git" && name != "node_modules"
        })
        .filter_map(Result::ok)
    {
        if entry.file_type().is_file() && entry.file_name() == "SKILL.md" {
            candidates.push(entry.into_path());
        }
    }
    println!("  found {} skill(s).", candidates.len());

    let target_root = repo_root()?.join(".claude").join("skills");
    fs::create_dir_all(&target_root).context("create .claude/skills/")?;

    let mut report: Vec<(String, Status, String)> = Vec::new();
    for skill_md in &candidates {
        let src_dir = skill_md.parent().ok_or_else(|| anyhow!("SKILL.md has no parent"))?;
        let name = skill_name(skill_md, src_dir)?;

        let dest = target_root.join(&name);
        let from_rel = src_dir
            .strip_prefix(workdir_path)
            .map(|p| p.to_string_lossy().into_owned())
            .unwrap_or_else(|_| src_dir.to_string_lossy().into_owned());

        if dest.exists() {
            if args.force {
                fs::remove_dir_all(&dest).context("remove existing skill dir")?;
                copy_dir(src_dir, &dest)?;
                report.push((name, Status::Overwritten, from_rel));
            } else {
                report.push((name, Status::Skipped, from_rel));
            }
        } else {
            copy_dir(src_dir, &dest)?;
            report.push((name, Status::Installed, from_rel));
        }
    }

    print_summary(&report);
    Ok(())
}

fn split_source(s: &str) -> Result<(String, String)> {
    let (repo, refname) = match s.split_once('#') {
        Some((r, f)) => (r.to_string(), f.to_string()),
        None => (s.to_string(), "HEAD".to_string()),
    };
    if !repo.contains('/') {
        bail!("Invalid source \"{s}\" \u{2014} expected <owner>/<repo>");
    }
    Ok((repo, refname))
}

fn skill_name(skill_md: &Path, src_dir: &Path) -> Result<String> {
    let text = fs::read_to_string(skill_md).context("read SKILL.md")?;
    if let Some(stripped) = text.strip_prefix("---\n") {
        if let Some(end) = stripped.find("\n---") {
            let fm_text = &stripped[..end];
            if let Ok(fm) = serde_yaml::from_str::<Frontmatter>(fm_text) {
                if let Some(n) = fm.name {
                    let trimmed = n.trim();
                    if !trimmed.is_empty() {
                        return Ok(trimmed.to_string());
                    }
                }
            }
        }
    }
    Ok(src_dir
        .file_name()
        .ok_or_else(|| anyhow!("skill dir has no name"))?
        .to_string_lossy()
        .into_owned())
}

fn copy_dir(src: &Path, dest: &Path) -> Result<()> {
    let opts = fs_extra::dir::CopyOptions {
        copy_inside: true,
        ..Default::default()
    };
    fs_extra::dir::copy(src, dest, &opts)
        .with_context(|| format!("copy {} -> {}", src.display(), dest.display()))?;
    Ok(())
}

fn repo_root() -> Result<PathBuf> {
    // Honor CARGO_WORKSPACE_DIR if set (cargo nightly), else `cargo metadata`-equivalent
    // via `git rev-parse --show-toplevel`, else current dir.
    if let Ok(dir) = env::var("CARGO_WORKSPACE_DIR") {
        return Ok(PathBuf::from(dir));
    }
    let out = Command::new("git")
        .args(["rev-parse", "--show-toplevel"])
        .output()
        .context("spawn git rev-parse")?;
    if out.status.success() {
        let path = String::from_utf8(out.stdout).context("git output utf-8")?;
        return Ok(PathBuf::from(path.trim()));
    }
    Ok(env::current_dir()?)
}

fn print_summary(report: &[(String, Status, String)]) {
    let name_w = report.iter().map(|(n, _, _)| n.len()).max().unwrap_or(4).max(4);
    let status_w = report
        .iter()
        .map(|(_, s, _)| s.as_str().len())
        .max()
        .unwrap_or(6)
        .max(6);

    println!();
    println!(
        "{:<name_w$}  {:<status_w$}  from",
        "name",
        "status",
        name_w = name_w,
        status_w = status_w
    );
    println!(
        "{:-<name_w$}  {:-<status_w$}  {}",
        "",
        "",
        "----",
        name_w = name_w,
        status_w = status_w
    );
    for (n, s, f) in report {
        println!(
            "{:<name_w$}  {:<status_w$}  {}",
            n,
            s.as_str(),
            f,
            name_w = name_w,
            status_w = status_w
        );
    }
    println!();
    let total = report.len();
    let installed = report.iter().filter(|(_, s, _)| matches!(s, Status::Installed)).count();
    let skipped = report.iter().filter(|(_, s, _)| matches!(s, Status::Skipped)).count();
    let overwritten = report
        .iter()
        .filter(|(_, s, _)| matches!(s, Status::Overwritten))
        .count();
    println!(
        "{total} skill(s) processed \u{2014} {installed} installed, {skipped} skipped, {overwritten} overwritten"
    );
}
